proposals/003-authentication-api.md

```diff
@@ -0,0 +1,516 @@
+        // Store the subject for use later
+        this.clientSubject = clientSubject;
+    }
+
+    TlsCredentials certificate(ServerCredentialContext context) {
+	if (this.clientSubject == null) {
+	    throw new IllegalStateException();
+	}
+	return certs.getOrDefault(
+	    this.clientSubject.getPrincipal(X500Principal.class),
+            context.defaultTlsCredentials());
+    }
+}
+
+class ExampleMTlsFilterFactory implements FilterFactory<, Map<X500Principal, TlsCredentials>> {
+    Map<X500Principal, TlsCredentials> certs;
+    public Map<X500Principal, TlsCredentials> initialize(FilterFactoryContext context, C config) {
+      certs = context.tlsCredentials(...)
+    }
+    
+    ExampleMTlsFilter createFilter(FilterFactoryContext context, I initializationData) {
+        return new ExampleMTlsFilter(certs)
+    }
+
+```
+
+An almost identical class could be used with the `ExampleSaslTerminatingFilter` from the previous section for SASL-to-TLS identity mapping.
+
+### Proposed API for learning about server authentication outcomes
+
+This is the mirror image of `ClientSubjectAware`, but for cases where the broker is using mTLS and/or a SASL mechanism that supports mutual authentication.
+It allows plugin behaviour to depend on the server's identity.
+
+```java
+package io.kroxylicious.proxy.authentication;
+
+import javax.security.auth.Subject;
+import javax.security.auth.login.LoginException;
+
+// Allows a plugin to opt in to being aware of broker-facing authentication outcomes
+interface ServerSubjectAware {
+    
+    /** 
+     * Called when the proxy authenticates, or reauthenticates with a server
+     * The given serverSubject may not have a principal for the server corresponding 
+     * to the authentication mechanism used if that authentication mechanism does not provide
+     * mutual authentication.
+     */
+    void onServerAuthentication(Subject serverSubject, ServerAuthenticationContext context);
+    
+    /**
+     * Called when the proxy fails authentication, or reauthentication, with a server
+     */
+    default void onServerAuthenticationFailure(LoginException exception, ServerAuthenticationContext context) { 
+    }
+}
+```
+
+```java
+package io.kroxylicious.proxy.authentication;
+
+public interface ServerAuthenticationContext {
+
+    /** 
+     * The subject that the proxy presented to the server.
+     */
+    Subject proxySubject();
+}
+```
+
+### Proposed API for initiating a SASL exchange with the broker 
+
+There are two mutually exclusive cases to consider for SASL initiation:
+
+1. That a SASL initiator plugin should perform SASL authentication _eagerly_, as soon as the underlying connection is ready. 
+This would be driven by the runtime following a successful TCP handshake on the client-to-proxy connection, and after `ClientSubjectAware` plugins have initially been called with any `X500Principal`, but before any SASL exchange on the client side.
```

Comment:
It would need to occur after the ApiVersions exchange, right?

Reply:
These sentences about how/when initiation occurs have dropped out. Closing as outdated.